### ISSUE TITLE
Where to place `countWire` utility function?

### ISSUE THREAD

**Issue by c-lewis** [karamaan/karamaan-opaleye#76]
Using the `count` aggregate function currently requires that it be provided a value:

``` haskell
exampleQ :: Query (Wire String, Wire Int64)
exampleQ = Q.aggregate (p3 (groupBy , count)) $ proc () -> do
    x <- exampleTable -< ()
    c <- toQueryArrDef $ showConstant (0 :: Int64)
    returnA -< (x , (c :: Wire Int64))
```

It might be possible to change `count`, so it generates the `Wire a` required by the `Aggregator`, making it equivalent to `count(*)`, but a `count` with the existing type signature will still be required in the situation in which you want to count the occurrence of non-null values in a column, equivalent to `count(col)`. (It might be nice to make this distinction explicit, but that's a different conversation).

Therefore, I propose adding a `countWire` utility function so everyone doesn't have to deal with the boilerplate above:

``` haskell
exampleQ :: Query (Wire String, Wire Int64)
exampleQ = Q.aggregate (p3 (groupBy , count)) $ proc () -> do
    x <- innerQ -< ()
    c <- countWire -< ()
    returnA -< (x , c)
```

The `Aggregate` module seems like the logical home for this utility function (that would require adding an `ExprArr` import).

Does the above seem reasonable/desirable?

**Comment by tomjaguarpaw** [comment]
Am I right in thinking

```
countWire :: QueryArr () (Wire Int64)
```

is just a constant `0`?

I can see what you mean about `count` needing to be provided a value being artificial.  Would you be happy with (instead of `countWire`)

```
countStar :: Aggregator () (Wire Int64)
```

Then you would just say

```
exampleQ :: Query (Wire String, Wire Int64)
exampleQ = Q.aggregate (p3 (groupBy , countStar)) $ proc () -> do
    x <- exampleTable -< ()
    returnA -< (x , ())
```

It's still slightly awkward, but better than having to pass an arbitrary value.

**Comment by c-lewis** [comment]
Your thinking in regards to `countWire` being a constant `0` is correct.

It wasn't clear to me if `countStar` would work, and I didn't spend any time on it because you will still need a `countColumn :: Aggregator (Wire a) (Wire Int64)` for the situation I referred to when you want to count the number of non-null values in a column (that's the only use I'm aware of, but there might be others). I'm not sure if this should really be `countNonNull :: Aggregator (Wire (Nullable a)) (Wire Int64)`, but I didn't see a reason to make a consumer be so specific, the former satisfies the later.

I prefer your `countStar`, but I would probably call it `countTuple` (seems clearer to me).

**Comment by tomjaguarpaw** [comment]
Having had a think about it, I think `countStar :: Aggregator () (Wire Int64)` is impossible, but I have an idea for a more flexible sort of aggregator that would make it possible.  I shall give some thought to it over the coming week.

In the meanwhile, I would be happy to add `zero :: Query (Wire Int64)` to `Operators2` if you don't want to keep it in your own project.

EDIT: corrected aggregator

**Comment by c-lewis** [comment]
I don't understand why you say `countColumn :: Aggregator (Wire a)
(Wire Int64)` is impossible? That's what `count` currently is.

Nothing has to be done now—give it some time.

**Comment by tomjaguarpaw** [comment]
Yes I meant `countStar :: Aggregator () (Wire Int64)` is impossible.  I updated my post but I don't think Github sends out updates by email.

**Comment by c-lewis** [comment]
Ah. That makes sense. I wasn't sure it would work either.

There should be a GitHub repository to which issue with GitHub can be
posted.

**Comment by tomjaguarpaw** [comment]
I added `countAll` with the desired functionality to `AggregatorSuper`.  The plan is to completely replace the old aggregators with `AggregatorSuper`s.
